flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java

```diff
@@ -17,95 +17,173 @@
+        this.committer = checkNotNull(committer);
+    }
+
+    @Override
+    public void setup(
+            StreamTask<?, ?> containingTask,
+            StreamConfig config,
+            Output<StreamRecord<CommittableMessage<CommT>>> output) {
+        super.setup(containingTask, config, output);
+        committableCollector = CommittableCollector.of(getRuntimeContext());
     }
 
     @Override
     public void initializeState(StateInitializationContext context) throws Exception {
         super.initializeState(context);
-        committerHandler.initializeState(context);
-        // try to re-commit recovered transactions as quickly as possible
-        commitRetrier.retryWithDelay();
+        committableCollectorState =
+                new SimpleVersionedListState<>(
+                        context.getOperatorStateStore()
+                                .getListState(STREAMING_COMMITTER_RAW_STATES_DESC),
+                        new CommittableCollectorSerializer<>(committableSerializer));
+
+        if (context.isRestored()) {
+            committableCollectorState.get().forEach(cc -> committableCollector.merge(cc));
+            lastCompletedCheckpointId = context.getRestoredCheckpointId().getAsLong();
+            // try to re-commit recovered transactions as quickly as possible
+            commitAndEmitCheckpoints();
+        }
     }
 
     @Override
     public void snapshotState(StateSnapshotContext context) throws Exception {
         super.snapshotState(context);
-        committerHandler.snapshotState(context);
+        // It is important to copy the collector to not mutate the state.
+        committableCollectorState.update(Collections.singletonList(committableCollector.copy()));
     }
 
     @Override
     public void endInput() throws Exception {
-        committerHandler.endOfInput();
-        commitRetrier.retryIndefinitely();
+        Collection<? extends CommittableManager<CommT>> endOfInputCommittables =
+                committableCollector.getEndOfInputCommittables();
+        // indicates batch
+        if (endOfInputCommittables != null) {
+            do {
+                for (CommittableManager<CommT> endOfInputCommittable : endOfInputCommittables) {
+                    commitAndEmit(endOfInputCommittable, false);
+                }
+            } while (!committableCollector.isFinished());
+        }
     }
 
     @Override
     public void notifyCheckpointComplete(long checkpointId) throws Exception {
         super.notifyCheckpointComplete(checkpointId);
-        committerHandler.notifyCheckpointCompleted(checkpointId);
-        commitRetrier.retryWithDelay();
+        lastCompletedCheckpointId = Math.max(lastCompletedCheckpointId, checkpointId);
+        commitAndEmitCheckpoints();
+    }
+
+    private void commitAndEmitCheckpoints() throws IOException, InterruptedException {
+        for (CheckpointCommittableManager<CommT> manager :
```

Comment:
I tried your suggestion unfortunately it feels like it complicates the solution. First I also need to pass the committer to the commit method
```
List<CommittableMessage> commit(int subtaskId, int totalSubtasks, long lastestCompletedCheckpointId, Committer<CommT> committer);
```
but the biggest problem is that I need to create the `CommittableSummary`s for every checkpoint so I cannot simply return a list of `CommittableMessage`s. The `CommittableManager` is the only object that holds all necessary information to build the correct summary.

Reply:
Thanks @fapaul for the explanation! I think perhaps `CommittableSummary` is also the subclass of the the `CommittableMessage`  and it should be able to be mixed in the returned list? And passing `committer` is indeed required, but it seems fine from my side~

Also may I have a double confirmation that It seems with the current implementation we'll emit `CommittableSummary` whenever we commit some committables instead of only when checkpointing? It seems to me it is to deal with the case that for unaligned checkpoint, the barrier might jump over all the summary and committable messages~ It should be also the reason we support receive multiple summary from one subtask for the same checkpoint, do I understand right~?

Reply:
> Thanks @fapaul for the explanation! I think perhaps `CommittableSummary` is also the subclass of the the `CommittableMessage` and it should be able to be mixed in the returned list? And passing `committer` is indeed required, but it seems fine from my side~
> 
Good point. To be honest I like the current attraction that we have one central place to coordinate the additions and removals of the `CommittableMessage`: the `CommittableCollector` (I am open for a different name). In addition, we have two different interfaces to actually interact with the committables one is `CommittableManager` that is relevant to commit committables in batch execution mode, and `CheckpointCommittableManager` is used in streaming execution mode to commit committables.



> Also may I have a double confirmation that It seems with the current implementation we'll emit `CommittableSummary` whenever we commit some committables instead of only when checkpointing? It seems to me it is to deal with the case that for unaligned checkpoint, the barrier might jump over all the summary and committable messages~ It should be also the reason we support receive multiple summary from one subtask for the same checkpoint, do I understand right~?